In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,root_mean_squared_error,mean_squared_error
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df=pd.read_csv(r'/content/adult_with_headers (1).csv')
df.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [5]:
df.describe()

,age,fnlwgt,education_num,capital_gain,capital_loss,hours_per_week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [6]:
print(df.isnull().sum())

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64


In [7]:
for col in df.select_dtypes(include='object').columns:
    if ' ?' in df[col].unique():
        print({df[col].value_counts()[' ?']})
        df[col] = df[col].replace(' ?', np.nan)

{np.int64(1836)}
{np.int64(1843)}
{np.int64(583)}


In [8]:
print(df.isnull().sum())

age                  0
workclass         1836
fnlwgt               0
education            0
education_num        0
marital_status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital_gain         0
capital_loss         0
hours_per_week       0
native_country     583
income               0
dtype: int64


In [11]:
for col in ['workclass', 'occupation', 'native_country']:
    if df[col].isnull().any():
        mode_value = df[col].mode()[0]
        df[col] = df[col].fillna(mode_value)

In [12]:
print(df.isnull().sum())

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64


Standard scaling is mostly used to features with continuous values, it will not have any fixed range between the values.  It is less affected by outliers. And it is used where the feature is like gassian value.  

Min-Max scaling used in a fixed range values for specific future.  It may affected by outliers.  When compared to Standard scaler it is less used mostly in some specific scenerio.

### One-Hot Encoding for Categorical Variables

One-Hot Encoding is a technique used to convert categorical variables into a numerical format that can be provided to machine learning algorithms. Each category is transformed into a new binary column (0 or 1).

We will apply this specifically to categorical columns that have less than 5 unique categories to avoid creating too many new features, which can lead to high dimensionality.

In [20]:
categorical_cols = df.select_dtypes(include='object').columns.tolist()
cols_to_encode = []
for col in categorical_cols:
    num_unique = df[col].nunique()
    if num_unique < 5:
        cols_to_encode.append(col)

In [21]:
if cols_to_encode:
    df_encoded = pd.get_dummies(df, columns=cols_to_encode, drop_first=False)
    display(df_encoded.head())
    df.shape
    df_encoded.shape
    df = df_encoded
else:
    print("no less than 5 unique")

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain,capital_loss,hours_per_week,native_country,sex_ Female,sex_ Male,income_ <=50K,income_ >50K
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,2174,0,40,United-States,False,True,True,False
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,0,0,13,United-States,False,True,True,False
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,0,0,40,United-States,False,True,True,False
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,0,0,40,United-States,False,True,True,False
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,0,0,40,Cuba,True,False,True,False


In [23]:
from sklearn.preprocessing import LabelEncoder
remaining_categorical_cols = df.select_dtypes(include='object').columns.tolist()
le = LabelEncoder()
for col in remaining_categorical_cols:
    df[col] = le.fit_transform(df[col])
display(df.head())
df.info()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,capital_gain,capital_loss,hours_per_week,native_country,sex_ Female,sex_ Male,income_ <=50K,income_ >50K
0,39,6,77516,9,13,4,0,1,4,2174,0,40,38,False,True,True,False
1,50,5,83311,9,13,2,3,0,4,0,0,13,38,False,True,True,False
2,38,3,215646,11,9,0,5,1,4,0,0,40,38,False,True,True,False
3,53,3,234721,1,7,2,5,0,2,0,0,40,38,False,True,True,False
4,28,3,338409,9,13,2,9,5,2,0,0,40,4,True,False,True,False


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   age             32561 non-null  int64
 1   workclass       32561 non-null  int64
 2   fnlwgt          32561 non-null  int64
 3   education       32561 non-null  int64
 4   education_num   32561 non-null  int64
 5   marital_status  32561 non-null  int64
 6   occupation      32561 non-null  int64
 7   relationship    32561 non-null  int64
 8   race            32561 non-null  int64
 9   capital_gain    32561 non-null  int64
 10  capital_loss    32561 non-null  int64
 11  hours_per_week  32561 non-null  int64
 12  native_country  32561 non-null  int64
 13  sex_ Female     32561 non-null  bool 
 14  sex_ Male       32561 non-null  bool 
 15  income_ <=50K   32561 non-null  bool 
 16  income_ >50K    32561 non-null  bool 
dtypes: bool(4), int64(13)
memory usage: 3.4 MB


OHE PRO:
It will give unique variables to the data.
the number increases with no. of data increases.
It doesn't give ranking to the data.
CON:
It will create more number of columns.

Label Pro:
It is used to give order for the data.
It is very simple that even single column is enough for this.
It is used of ordinal data.
It uses less memory.
CON:
It may create false order for nominal data.

In [24]:
df['Total_Capital_Movement'] = df['capital_gain'] - df['capital_loss']
display(df[['capital_gain', 'capital_loss', 'Total_Capital_Movement']].head())

,capital_gain,capital_loss,Total_Capital_Movement
0,2174,0,2174
1,0,0,0
2,0,0,0
3,0,0,0
4,0,0,0


In [26]:
df['Hours_Per_Age'] = df['hours_per_week'] / df['age']
display(df[['age', 'hours_per_week', 'Hours_Per_Age']].head())

,age,hours_per_week,Hours_Per_Age
0,39,40,1.025641
1,50,13,0.260000
2,38,40,1.052632
3,53,40,0.754717
4,28,40,1.428571


In [27]:
df.skew()

,0
age,0.558743
workclass,0.076469
fnlwgt,1.446980
education,-0.934042
education_num,-0.311676
marital_status,-0.013508
occupation,0.000565
relationship,0.786818
race,-2.435386
capital_gain,11.953848


In [31]:
df['capital_gain_log'] = np.log1p(df['capital_gain'])
display(df[['capital_gain', 'capital_gain_log']].head())

print(f"{df['capital_gain'].skew():.4f}")
print(f"{df['capital_gain_log'].skew():.4f}")

,capital_gain,capital_gain_log
0,2174,7.684784
1,0,0.000000
2,0,0.000000
3,0,0.000000
4,0,0.000000


11.9538
3.0961


To address the high positive skewness (11.95) observed in the capital_gain feature, a log transformation using np.log1p was applied. This transformation significantly reduced the skewness to 3.0961, making the feature's distribution more symmetric and thus more suitable for machine learning models that often assume normally distributed input features.